### PR TITLE
Fix semicolon enforcement

### PR DESCRIPTION
**Pull request by Saalvage**
Woo, C++11!

## More-or-Less-Fun Fact
I would like to draw attention to the peculiarity that `DOCTEST_TEST_SUITE_END` poses. It is usable both as a function-like macro and as a regular macro. This is because in order to enforce the macro invocation be followed by a semicolon it has a typedef of the following form at the end: `typedef int DOCTEST_INTERNAL_STH`. Now, this is just adds a simple alias for int, doesn't it? (It does!)
But now watch what happens when we append parentheses: `typedef int DOCTEST_INTERNAL_STH()`. Now, I wouldn't blame anyone for believing this to be invalid syntax, but it actually introduces `DOCTEST_INTERNAL_STH` as an alias, for a ***function*** returning int and taking no parameters! Now isn't that crazy?

## TLDR
All macros that use the existing form of semicolon enforcement outside of functions may have their invocation followed by a set of parantheses inside which type parameters can be placed.
`DOCTEST_TEST_SUITE_END` is a special case because it is the only affected macro which is itself not function-like. This means backwards compatibility with both invocations should preferably be upheld for it, and it alone (who would ever do `DOCTEST_TEST_SUITE_BEGIN(...)()`?). Maintaining the old syntax in this case should not be a big deal either, as I don't think a test suite can ever be ended locally. 

## Related issue
#577

### CODE REVIEW

**Comment by onqtam** [comment]
In the future some compiler will complain that this static assert is unnecessary :P
